### PR TITLE
feat: quiz backend part completed

### PR DESCRIPTION
**Pull request by Jaynil1611**
- [Quiz Backend Live](https://dietify-quiz-backend.vercel.app/)

### CODE REVIEW

File: backend/package.json [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  "scripts": {
+    "start": "node index.js",
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "keywords": [],
+  "author": "",
+  "license": "ISC",
+  "dependencies": {
+    "bcrypt": "^5.0.1",
+    "body-parser": "^1.19.0",
```
Comment (Aravindan07) [inline backend/package.json:15]: You don't need to install body-parser separately it is pre-built in the express package.

File: backend/controllers/login.controller.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+      );
+      const { firstname, lastname } = user;
+      return res
+        .status(201)
+        .json({ success: true, token, firstname, lastname });
+    }
+
+    res
+      .status(401)
+      .json({ success: false, errorMessage: "Passwords don't match" });
```
Comment (Aravindan07) [inline backend/controllers/login.controller.js:32]: These error messages are not consumed by the UI I guess. You are showing hard-coded error messages why not use these messages. Not a backend fix though, just mentioning the issue.

File: backend/controllers/login.controller.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        .json({ success: false, errorMessage: "User doesn't exist!" });
+    }
+
+    const isPasswordMatch = await bcrypt.compare(password, user.password);
+    if (isPasswordMatch) {
+      const token = jwt.sign(
+        {
+          userId: user._id,
+          expiry: "24h",
+        },
```
Comment (Aravindan07) [inline backend/controllers/login.controller.js:21]: Why you created a separate file for login when you already have a user.controller.js file. You put the logic of signup there why not login too?

File: backend/middlewares/paramHandler.js [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+      return res
+        .status(400)
+        .json({ success: false, message: "Quiz not found!" });
+    }
+    req.quiz = quiz;
+    next();
+  } catch (error) {
+    res.status(400).json({
+      success: false,
+      message: "Couldn't retrieve the quiz",
```
Comment (Aravindan07) [inline backend/middlewares/paramHandler.js:17]: Couldn't retrieve the quiz sounds like a server issue. If it is then change the status code to 5XX.

File: backend/middlewares/paramHandler.js [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+      errorMessage: error.message,
+    });
```
Comment (Aravindan07) [inline backend/middlewares/paramHandler.js:41]: Same here too check the status code

File: backend/controllers/quizAttempt.controller.js [hunk 6]
```diff
@@ -0,0 +1,94 @@
+const { QuizAttempt } = require("../models/quizAttempt.model");
+const { extend } = require("lodash");
+const { getRequiredDataFromQuiz } = require("../utils/quizAttemptUtils");
+
+const getAllQuizAttempts = async (req, res, next) => {
+  const { userId } = req;
+  const quizzes = QuizAttempt.find({ userId });
+  res.status(201).json({ success: true, quizzes });
```
Comment (Aravindan07) [inline backend/controllers/quizAttempt.controller.js:8]: It's a get request so the status code will be 200.

File: backend/controllers/quizAttempt.controller.js [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+};
+
+const updateQuizAttempt = async (req, res, next) => {
+  try {
+    const quizUpdates = req.body;
+    let { quiz } = req;
+    console.log(quizUpdates);
+    quiz = extend(quiz, getRequiredDataFromQuiz(quizUpdates));
+    quiz = await quiz.save();
+    console.log({ quiz });
```
Comment (Aravindan07) [inline backend/controllers/quizAttempt.controller.js:49]: Remove logs

File: backend/utils/createUser.js [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+      username: "dieitifyQuiz",
+      password: "quizify1234",
+    });
+    await user.save();
+  } catch (error) {
+    console.error("Error while registering user", error.message);
+  }
+}
+
+//userId: 60ae411072e44c2e44da4814
```
Comment (Aravindan07) [inline backend/utils/createUser.js:15]: remove comments

Comment (Aravindan07) [inline backend/utils/createUser.js:15]: You are not using this function I think. If you are not using then remove it.

**Review by Aravindan07 — CHANGES_REQUESTED** [review]
Great work Jaynil with the quiz app. I have included some minor changes.

**Review by aneenajohn — CHANGES_REQUESTED** [review]
Good work, Jaynil!
Its a pleasure to look through such well structured code. I had suggested some minor changes

**Pull request merged**